nautobot/vpn/models.py

```diff
@@ -0,0 +1,522 @@
+        on_delete=models.SET_NULL,
+        related_name="vpn_tunnel_endpoints_src_ip",
+        blank=True,
+        null=True,
+        verbose_name="Source IP Address",
+        help_text="Mutually Exclusive with Source FQDN.",
+    )
+    source_fqdn = models.CharField(
+        max_length=CHARFIELD_MAX_LENGTH,
+        blank=True,
+        verbose_name="Source FQDN",
+        help_text="Mutually Exclusive with Source IP Address",
+    )
+    tunnel_interface = models.OneToOneField(
+        to="dcim.Interface",
+        on_delete=models.SET_NULL,
+        related_name="vpn_tunnel_endpoints_tunnel",
+        blank=True,
+        null=True,
+        verbose_name="Tunnel Interface",
+    )
+    vpn_profile = models.ForeignKey(
+        to="vpn.VPNProfile",
+        on_delete=models.PROTECT,
+        related_name="vpn_tunnel_endpoints",
+        blank=True,
+        null=True,
+        verbose_name="VPN Profile",
+    )
+    role = RoleField(blank=True, null=True)
+    protected_prefixes = models.ManyToManyField(
+        to="ipam.Prefix",
+        related_name="vpn_tunnel_endpoints",
+        blank=True,
+        verbose_name="Protected Prefixes",
+    )
+    protected_prefixes_dg = models.ManyToManyField(
+        to="extras.DynamicGroup",
+        related_name="vpn_tunnel_endpoints",
+        blank=True,
+        verbose_name="Protected Prefixes Dynamic Group",
+    )
+    tenant = models.ForeignKey(
+        to="tenancy.Tenant",
+        on_delete=models.SET_NULL,
+        related_name="vpn_tunnel_endpoints",
+        blank=True,
+        null=True,
+    )
+
+    clone_fields = [
+        "vpn_profile",
+        "device",
+        "source_interface",
+        "source_ipaddress",
+        "source_fqdn",
+        "tunnel_interface",
+        "protected_prefixes",
+        "protected_prefixes_dg",
+    ]
+
+    class Meta:
+        """Meta class for VPNTunnelEndpoint."""
+
+        ordering = ("name",)
+        verbose_name = "VPN Tunnel Endpoint"
+
+    def _name(self):
+        """Dynamic name field."""
+        if self.source_interface:
+            device_intf = f"{self.source_interface.device.name} {self.source_interface.name}"
+            if self.source_ipaddress:
+                return f"{device_intf} ({self.source_ipaddress.address})"
+            return device_intf
+        return self.source_fqdn
+
+    def __str__(self):
+        """Stringify instance."""
+        return self.name
+
+    def clean(self):
+        if self.source_ipaddress and self.source_fqdn:
+            raise ValidationError("Source IP Address and Source FQDN are mutually exclusive fields. Select only one.")
+        if not any([self.source_interface, self.source_ipaddress, self.source_fqdn]):
+            raise ValidationError("Source Interface or Source IP Address or Source FQDN Is required.")
+        if self.source_interface and not self.source_interface.device:
+            raise ValidationError("Source Interface must belong to a device.")
```

Comment:
Is this intentionally disallowing use of module interfaces?

Reply:
That's an obvious miss. In the model `clean()` we should then check `interface.device or interface.module` then.